### ISSUE TITLE
akka stream TCP does not close properly

### ISSUE THREAD

**Issue by RustedBones** [akka/akka#22163]
I am using a TCP outgoing connection stream connected to 2 actors: source ~> tcp ~> sink.

If the remote hosts closes the connection, the sink actor receives the `onCompleteMessage` as expected. I expect at this point I can safely stop the sink actor. According to the doc, I should not care about the source actor:
> The actor will be stopped when the stream is completed

However, when running wireshark and exterminating the TCP message send the flow is the following:

server -> client: [FIN, ACK]
client -> server: [RST, ACK]

I was expecting a normal termination of the connection:

server -> client: [FIN, ACK]
client -> server: [FIN, ACK]

Sounds like `RST` is used when the stream is canceled from downstream. So I did the following:
Instead of stopping the sink actor when receiving the `onCompleteMessage`, I tried to send `Status.Success(NotUsed)` to the source before stopping. I expected the message to go to the dead-letter, but instead, it  solves the problem.

TCP flow materialized with an `Sink.actorRef` does not respect the akka-stream specification:

> When the stream is completed successfully the given `onCompleteMessage` will be sent to the destination actor.

In the above case, we got the `onCompleteMessage` but the stream is still alive.

**Comment by RustedBones** [comment]
After more testing, I realized that `client -> server: [RST, ACK]` is not sent when the sink actor stops, but when the tests ends (so when the actor system terminates).

Stopping the sink actor does not trigger stream cancellation. The server still thinks the connection is alive and keeps waiting for input data.

The architecture I have is perhaps the cause but that should technically not be an issue:
```scala
class Actor(host: String, port: Int, bufferSize: Int) {
  case object Completed

  val src = Source.actorRef[ByteString](bufferSize, overflowStrategy = OverflowStrategy.fail)
  val flow = Tcp().outgoingConnection(remoteAddress  = new InetSocketAddress(host, port))
  val sink = Sink.actorRef(self, Completed)
  val (stream, connection) = src.viaMat(flow)(Keep.both).to(sink).run()
}
```

**Comment by nktpro** [comment]
@RustedBones I ran into this exact issue, and I'm doing exactly what you did: actorRefs for both Source and Sink, via TCP outgoing connection flow. I found it out the hard way with the service blowing up due to thousands of open sockets, even after the streams already completed and the Sink actors emitting the Completed message.

I strongly agree this is a bug. We should not have to manually send `Status.Success` to the source actor to close the socket.

**Comment by raboof** [comment]
I think the thing here is that TCP allows connections to be 'half-open': when the one side (here the server) sends a FIN, it signals that it won't send any more data (here: the TCP GraphStage correctly completes the stream towards the Sink), but it's still legal to keep sending data in the other direction (here: to the server) - at least from the TCP perspective. Most sane protocols will of course describe that the client should also shut down its side of the socket - AFAIK half-open is rarely used, and even then usually in the 'other direction'.

I'm not entirely sure what the most elegant way would be to make sure the 'writing' side of the stream completes, too, though - that seems like a common requirement that should be either trivial or even default behavior.

**Comment by RustedBones** [comment]
@raboof so the `halfClose` parameter to the [Tcp.outgoingConnection](http://doc.akka.io/api/akka/2.5/akka/stream/javadsl/Tcp.html) only works on the client (local) bounds, not on the remote bounds. That can make sense, but it is not obvious.

When Tls is used on top of Tcp, the stream will close properly and the ports do not stay open: Tls does not support halfClose. I think that could be nice to follow the same default behaviour when using Tcp only.

@nktpro I think there is another thing that can be done to solve the problem with the above design. Instead of sending a `Status.Success`  we may use a local `ActorMaterializer` in the Sink actor.  That should work if you stop this actor when the stream completes.

**Comment by raboof** [comment]
@RustedBones right, the `halfClose` parameter is 'the other way around', allowing the server to keep writing after our client has finished writing. The situation in the example above is that the server finished writing, but our clients' stream towards the server is still kept functional.

The underlying TCP system already has a `keepOpenOnPeerClosed` flag for connections, I think it might make sense to expose that for outgoing connections (next to `halfClose`). I'm having a hard time deciding what the default should be though: on the one hand 'leaking' open streams is bad, on the other hand withholding data from the server when it's eager in closing its side of the connection seems dangerous, too.

I'd like someone more familiar with that codebase to weigh in :).

**Comment by jrudolph** [comment]
We have been discussing the same problem with WebSockets as well. I agree the current behavior is problematic as it will the leak the source side of a client stream if a server closes the connection first.

I'm a bit worried about the proliferation of extra arguments to the API. In the most cases you don't want half-closed connections but still want to close connections gracefully. So, the best default setting would be a connection that doesn't support half-open connections but has a timeout that after one side was closed will close the other side. Usually that works well because (TCP) protocols are usually not independently bidirectional but have a defined way of closing a connection (both sides agree on closing the connection after a certain packet, `Connection: close` in HTTP/1.1, `GOAWAY` in HTTP/2, `close_notify` in TLS). That timeout would then control both sides. You could set it to infinite to disable it if you really want full control over the bidirectional connection if that's what you want but then you need to care for leaks yourself.

WDYT about that idea, @raboof?
